microsoft/testsuites/cvm/cvm_boot.py

```diff
@@ -0,0 +1,146 @@
+        )
+
+        assert_that(root_partition, "Cannot locate root partition").is_not_none()
+        assert isinstance(root_partition, PartitionInfo)
+        assert_that(root_partition.fstype).is_equal_to("crypto_LUKS")
+
+    @TestCaseMetadata(
+        description="""This test case verifies that a CVM can still boot if any boot
+        component is upgraded.
+
+        Steps:
+        1. On first boot, check current PCR values for PCR4 and PCR7
+        2. Get current boot components versions (e.g. shim, grub, systemd-boot, uki)
+        3. Run a package upgrade to update boot components
+        4. Get new boot components versions to see if anything has changed
+        5. Reboot the CVM, make sure the CVM can boot up again
+        6. PCR4 should change if any of the boot components is upgraded
+        7. PCR7 may change (for example, if a signing certificate is changed)
+        """,
+        priority=1,
+        requirement=simple_requirement(
+            environment_status=EnvironmentStatus.Connected,
+            supported_features=[CvmEnabled()],
+            supported_platform_type=[AZURE],
+        ),
+    )
+    def verify_boot_success_after_component_upgrade(
+        self, log: Logger, node: RemoteNode, log_path: Path
+    ) -> None:
+        posix_os: Posix = cast(Posix, node.os)
+        # First boot - no package upgrade has been performed
+        # Check PCR values (PCR4, PCR7)
+        pcrs_before_reboot = node.tools[Tpm2].pcrread(pcrs=[4, 7])
+
+        # - Get current boot components versions (shim, systemd-boot, kernel-uki)
+        boot_components = ["shim", "systemd-boot", "kernel-uki"]
+        boot_components_versions: Dict[str, str] = dict()
+        for pkg in boot_components:
+            pkg_version = posix_os.get_package_information(pkg, use_cached=False)
+            boot_components_versions[pkg] = pkg_version.version_str
+
+        # Upgrade boot components
+        posix_os.update_packages(boot_components)
+
+        # Get new boot components versions
+        boot_components_new_versions: Dict[str, str] = dict()
+        for pkg in boot_components:
+            pkg_version = posix_os.get_package_information(pkg, use_cached=False)
+            boot_components_new_versions[pkg] = pkg_version.version_str
+
+        # Reboot
+        node.reboot()
+
+        # VM is up again
+        pcrs_after_reboot = node.tools[Tpm2].pcrread(pcrs=[4, 7])
+        boot_component_changed = (
+            boot_components_versions != boot_components_new_versions
+        )
+
+        # - PCR4 should change if any of the boot components is upgraded
+        # - PCR7 may change if a signing cert is changed
+        if boot_component_changed:
+            assert_that(
+                pcrs_after_reboot[4],
+                "PCR4 value is still the same even though a boot component changed",
+            ).is_not_equal_to(pcrs_before_reboot[4])
+            assert_warn(
```

Comment:
Why is it a warning? The warning means it's not important and should be an info level bug. If it's important, make the test case fail. The middle area is not a good practice.

Reply:
updated to only use `log.info`

Reply:
Please change it to assert_that or info level log.

Reply:
I forgot to remove old code. Corrected